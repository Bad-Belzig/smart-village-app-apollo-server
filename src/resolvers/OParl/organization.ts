import { FilterQuery } from 'mongoose';

import {
  Body,
  Consultation,
  IOrganization,
  IOrganizationSchema,
  Location,
  Meeting,
  Membership,
  Organization,
} from '../../models';
import { getPaginatedEntriesByIds } from '../resolverHelpers';

export const organizationResolvers = {
  Query: {
    oParlOrganizations: (
      _,
      args: { externalIds?: string[]; keyword?: string[] },
    ) => {
      const filter: FilterQuery<IOrganizationSchema> = {};

      if (args.keyword?.length) filter.keyword = { $all: args.keyword };

      if (args.externalIds) filter.externalId = { $in: args.externalIds };
      return Organization.find(filter);
    },
  },
  OParlOrganization: {
    body: (parent: IOrganization) => Body.findOne({ externalId: parent.body }),
    consultation: async (
      parent: IOrganization,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Consultation,
        parent.consultation,
        args.offset,
        args.pageSize,
      ),
    meeting: async (
      parent: IOrganization,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Meeting,
        parent.meeting,
        args.offset,
        args.pageSize,
      ),
    membership: async (
      parent: IOrganization,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Membership,
        parent.membership,
        args.offset,
        args.pageSize,
      ),
    subOrganizationOf: (args: IOrganization) =>
      Organization.findOne({ externalId: args.subOrganizationOf }),
    location: (args: IOrganization) =>
      Location.findOne({ externalId: args.location }),
  },
};
