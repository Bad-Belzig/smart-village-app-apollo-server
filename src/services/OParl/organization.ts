import { fetch } from 'apollo-env';

import { Organization } from '../../models';
import { parseOrganization } from '../../parser';
import { fetchPaginatedOParlList } from './paginatedList';

// import flow
// 1. fetch json
// 2. create db object
//   2.1. validation necessary for parsing (done in parser)
//   2.2. parse json
//   2.3. create object
// 3. validate db object
// 4. queue new related objects
// 5. save to db

export const importOrganization = async (
  url: string,
  createdSince?: Date,
  createdUntil?: Date,
) => {
  try {
    const response = await fetch(url);

    if (response.ok) {
      const json = await response.json();

      // TODO: check for previous existence

      if (json?.meeting) {
        json.meeting = await fetchPaginatedOParlList(
          json.meeting,
          createdSince,
          createdUntil,
        );
      }

      const lt = new Organization(parseOrganization(json));

      await lt.validate();

      // TODO: check for person and organization and meetings to import

      return lt.save();
    } else {
      throw new Error(
        `Error while fetching Organization from : ${url}! \n response.status: ${response.status}`,
      );
    }
  } catch (e) {
    console.log(e);
  }
};
