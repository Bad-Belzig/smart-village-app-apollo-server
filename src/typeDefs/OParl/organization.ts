import { gql } from 'apollo-server-core';

import { oParlBase } from './oParlBase';

export const organizationGQL = gql`
  extend type Query {
    oParlOrganizations(
      externalIds: [String!]
      keyword: [String!]
    ): [OParlOrganization!]
  }

  type OParlOrganization {
    body: OParlBody
    classification: String
    endDate: Date
    externalBody: String
    location: OParlLocation
    meeting(offset: Int, pageSize: Int): [OParlMeeting!]
    membership(offset: Int, pageSize: Int): [OParlMembership!]
    consultation(offset: Int, pageSize: Int): [OParlConsultation!]
    name: String
    organizationType: String
    post: [String!]
    shortName: String
    startDate: Date
    subOrganizationOf: OParlOrganization
    website: String
    ${oParlBase}
  }
`;
