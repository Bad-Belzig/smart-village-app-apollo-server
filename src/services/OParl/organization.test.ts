import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../helpers';
import { importOrganization } from './organization';

const helperFuncs = setupTestDataBase('test_organization_import');

// TODO: change test url to something within our control
const testUrl =
  'https://oparl.politik-bei-uns.de/organization/5a73013bf24bb7ed5127a7df';
const testUrl2 =
  'https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/organization/1-3';

describe('importing an organization', () => {
  beforeAll(helperFuncs.beforeAll);

  it(
    'imports properly for 1.0',
    basicImportTest(
      importOrganization,
      testUrl,
      new Date('2017-06-01'),
      new Date('2018-06-01'),
    ),
  );

  it('imports properly for 1.1', basicImportTest(importOrganization, testUrl2));

  it(
    'import returns undefined on faulty url',
    invalidUrlTest(importOrganization),
  );

  afterAll(helperFuncs.afterAll);
});
