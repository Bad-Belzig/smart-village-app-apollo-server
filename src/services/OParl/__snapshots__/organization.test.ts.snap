// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing an organization imports properly for 1.0 1`] = `
Object {
  "__v": 0,
  "body": "https://oparl.politik-bei-uns.de/body/5a73013af24bb7ed5127a72a",
  "classification": "Ausschuss",
  "consultation": Array [],
  "created": 2016-10-06T10:00:00.000Z,
  "externalId": "https://oparl.politik-bei-uns.de/organization/5a73013bf24bb7ed5127a7df",
  "keyword": Array [],
  "location": "https://oparl.politik-bei-uns.de/location/5a73013bf24bb7ed5127a7b5",
  "meeting": Array [
    "https://oparl.politik-bei-uns.de/meeting/5a730150f24bb7ed5127d6eb",
    "https://oparl.politik-bei-uns.de/meeting/5a730175f24bb7ed5127eac8",
    "https://oparl.politik-bei-uns.de/meeting/5a73017ff24bb7ed5127f0d7",
    "https://oparl.politik-bei-uns.de/meeting/5a730183f24bb7ed5127f330",
  ],
  "membership": Array [
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7b8",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7bb",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7be",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7c1",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7c4",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7c7",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7ca",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7cd",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7d0",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7d3",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7d6",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7d9",
    "https://oparl.politik-bei-uns.de/membership/5a73013bf24bb7ed5127a7dc",
  ],
  "modified": 2018-01-23T10:59:02.000Z,
  "name": "Betriebsausschuss Abfallwirtschaft",
  "organizationType": "Gremium",
  "post": Array [],
  "shortName": "BA",
  "startDate": 2016-10-06T00:00:00.000Z,
  "type": "https://schema.oparl.org/1.0/Organization",
}
`;

exports[`importing an organization imports properly for 1.1 1`] = `
Object {
  "__v": 0,
  "body": "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1",
  "classification": "Ausschüsse",
  "consultation": Array [],
  "created": 2015-03-18T16:05:27.000Z,
  "externalId": "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/organization/1-3",
  "keyword": Array [],
  "location": "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/location/1-3",
  "meeting": Array [],
  "membership": Array [
    "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/membership/187-1-3-94",
    "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/membership/205-1-3-53",
    "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/membership/8-1-3-59",
    "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/membership/195-1-3-77",
    "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/membership/45-1-3-80",
    "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/membership/206-1-3-183",
    "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/membership/191-1-3-187",
  ],
  "modified": 2017-05-04T22:00:00.000Z,
  "name": "Rechnungsprüfungsausschuss",
  "organizationType": "Gremium",
  "post": Array [
    "Vorsitzende/r",
    "stellv. Vorsitzende/r",
    "Mitglied",
  ],
  "shortName": "RPA",
  "startDate": 1950-01-01T00:00:00.000Z,
  "type": "https://schema.oparl.org/1.1/Organization",
  "web": "https://sdnetrim.kdvz-frechen.de/rim4883/gremien/?__=UGhVM0hpd2NXNFdFcExjZdagC1p12Ukom34sARBPmlg",
  "website": "",
}
`;
